File: src/encoder_manager_node.cpp
#include "umrt-serial-cam-ros/encoder_manager_node.hpp"
#include <functional> // For std::bind

namespace umrt_serial_cam_ros
{

/*
 * Public Functions 
 */

/**
 * Constructor
 */
EncoderManager::EncoderManager():Node("encoder_manager")
{  
    //  Info
    RCLCPP_INFO(this->get_logger(), "Encoder Manager Node Started");

    // Initialize encoder configurations
    encoder_configs_["rover_cam0"] = {
        "rover",
        "/rover/image_raw",
        "/rover",
        "h264_nvenc",
        "31",
        "p1",
        "ll",
        "2000000",
        "15"};

    encoder_configs_["arm_cam0"] = {
        "arm0",
        "/arm0/image_raw",
        "/arm0",
        "h264_nvenc",
        "31",
        "p1",
        "ll",
        "2000000",
        "15"};

    encoder_configs_["arm_cam1"] = {
        "arm1",
        "/arm1/image_raw",
        "/arm1",
        "h264_nvenc",
        "31",
        "p1",
        "ll",
        "2000000",
        "15"};

}   //  Constructor

/**
 * Destructor
 */
EncoderManager::~EncoderManager()
{
    // Stop any running encoders on shutdown
    for (const auto &encoder : encoder_status_)
    {
        const std::string &cam_name = encoder.first;
        if (encoder.second)
        {
            RCLCPP_INFO(this->get_logger(), "Stopping encoder for %s during shutdown.", cam_name.c_str());
            stopEncoder(cam_name);
        }
    }

    RCLCPP_INFO(this->get_logger(), "Encoder Manager Node Shutting Down.");
}   //  Destructor

/**
 * init Method
 */
void EncoderManager::init()
{

    //  Create services and setup the encoder for each camera
    for (const auto &camera : encoder_configs_)
    {
        const std::string &cam_name = camera.first;
        std::string service_name = "/" + cam_name + "/bool";

        encoder_status_[cam_name] = false;

        //  Use std::bind to pass the camera_namespace
        control_services_[cam_name] = this->create_service<std_srvs::srv::SetBool>(
            service_name, 
            std::bind(&EncoderManager::encoderControlCallback, this,
                      std::placeholders::_1, std::placeholders::_2, cam_name)
        );

        RCLCPP_INFO(this->get_logger(), "Created service: %s for %s", service_name.c_str(), cam_name.c_str());
    
        //  Call the setup function to create the encoders (publishers and subscribers)
        setupEncoder(cam_name);
    }
}

/*
 * Private Functions 
 */

/**
 * setupEncoder - Will start all cameras on node startup, given each camera_namespace, 
 * this is to get rid of the start and kill of encoders
 */
void EncoderManager::setupEncoder(const std::string &camera_namespace)
{
    //  Try to setup encoder
    try
    {

        // Get config for a given camera
        const EncoderConfig &config = encoder_configs_.at(camera_namespace);

        // Create a new ImageTransport instance for this camera
        image_transports_[camera_namespace] = std::make_shared<image_transport::ImageTransport>(shared_from_this());

        // The second argument specifies the transport plugin. "ffmpeg" is what we want.
        image_transport::ImageTransport& it = *image_transports_[camera_namespace];

        // Create the image_transport publisher configured for ffmpeg encoding
        // Publisher
        // QoS
        rclcpp::QoS publisher_qos(1);
        publisher_qos.reliable();
        publisher_qos.durability_volatile();
        rclcpp::PublisherOptions pub_options;

        // Transport Hints - for the ffmpeg image transport plugin
        // image_transport::TransportHints pub_transport_hints(shared_from_this().get(), "ffmpeg");

        encoder_publishers_[camera_namespace] = it.advertise(
            config.out_ffmpeg_topic,
            publisher_qos.get_rmw_qos_profile()
        ); // Publisher

        RCLCPP_INFO(this->get_logger(), "Advertised encoder topic: %s using encoder: %s",
            config.out_ffmpeg_topic.c_str(), config.encoding.c_str());

        // Create the image_transport subscriber to listen to the raw image topic
        // Subscriber
        // QoS
        rclcpp::QoS subscriber_qos(1);
        subscriber_qos.reliable();  //  I CANT MAKE THE DECODER BEST EFFORT AAAAAAAA
        subscriber_qos.durability_volatile();
        rclcpp::SubscriptionOptions sub_options;

        // Transport Hints - for the ffmpeg image transport plugin
        image_transport::TransportHints sub_transport_hints(shared_from_this().get(), "raw");

        raw_image_subscribers_[camera_namespace] = it.subscribe(
            config.in_raw_topic,
            subscriber_qos.get_rmw_qos_profile(),
            [this, camera_namespace](const sensor_msgs::msg::Image::ConstSharedPtr &msg) 
            {
                this->rawImageCallback(msg, camera_namespace);
            },
            shared_from_this(),
            &sub_transport_hints,
            sub_options
        ); // Subscriber
                
        // ROS INFO
        RCLCPP_INFO(this->get_logger(), "Encoder for %s started. Subscribing to %s, Publishing to %s (ffmpeg).",
                    camera_namespace.c_str(), config.in_raw_topic.c_str(), config.out_ffmpeg_topic.c_str());

        //  Set the encoder status to true
        encoder_status_[camera_namespace] = true;
        
    }
    //  Catch exception if try fails
    catch (const std::exception &e)
    {
        RCLCPP_ERROR(this->get_logger(), "Failed to start encoder %s: %s", camera_namespace.c_str(), e.what());
        encoder_status_[camera_namespace] = false;
    }

}   //  setupEncoder

 /**
 * encoderControlCallback - Backbone essentially, will get the bool service and check to start or stop the encoders.
 */
void EncoderManager::encoderControlCallback(
    const std::shared_ptr<std_srvs::srv::SetBool::Request> request,
    std::shared_ptr<std_srvs::srv::SetBool::Response> response,
    const std::string &camera_namespace)
{
    //  Get the SetBool Service request and then find out if it is start or stop
    bool command_bool = request->data;
    std::string command = command_bool ? "start" : "stop"; // Maybe we can set this to an int like 0 or 1 

    RCLCPP_INFO(this->get_logger(), "Received command: %s for %s", command.c_str(), camera_namespace.c_str());

    //  Get current encoder status
    bool current_encoder_status = encoder_status_[camera_namespace];

    //  Check if command is valid
    //  Start Command - To start encoder
    if (command == "start")
    {
        //  Check if encoder for that namespace is already running 
        //  If it is running, throw a warning
        if (current_encoder_status)
        {
            response->success = false;
            response->message = "Encoder " + camera_namespace + " is already running.";
            RCLCPP_WARN(this->get_logger(), response->message.c_str()); 
        }
        //  Else if its not running, attempt to start 
        else 
        {   
            startEncoder(camera_namespace);
            response->success = true;
            response->message = "Encoder " + camera_namespace + " started publishing.";
        }
    }
    //  Stop Command - To stop an encoder
    else if (command == "stop")
    {
        //  Check if encoder for that namespace has stopped
        //  If it is stopped, thrown a warning
        if (!current_encoder_status)
        {
            response->success = false;
            response->message = "Encoder " + camera_namespace + " is already stopped.";
            RCLCPP_WARN(this->get_logger(), response->message.c_str());
        }
        //  Else if it is not stopped, attempt to stop
        else 
        {
            stopEncoder(camera_namespace);
            response->success = true;
            response->message = "Encoder " + camera_namespace + " stopped publishing.";
        }
    }
    //  Invalid Command
    else 
    {
        response->success = false;
        response->message = "Unknown command: " + command + ". Use 'start' or 'stop'";
        RCLCPP_ERROR(this->get_logger(), response->message.c_str());
    }

}   //  encoderControlCallback

/**
 * rawImageCallback - Function is to forward the raw images to the image_transport publisher
 *                      Hopefully the ffmpeg_image_transport will the the encoding
 */
void EncoderManager::rawImageCallback(
    const sensor_msgs::msg::Image::ConstSharedPtr &msg,
    const std::string &camera_namespace)
{
    if (encoder_status_[camera_namespace])
    {
        encoder_publishers_[camera_namespace].publish(msg);
    }
}   //  rawImageCallback

/**
 * startEncoder Function - This will start an encoder which is essentially a republish
 */
bool EncoderManager::startEncoder(const std::string &camera_namespace)
{
    encoder_status_[camera_namespace] = true;
    RCLCPP_INFO(this->get_logger(), "Publishing started for node %s", camera_namespace.c_str());
}   //  startEncoder

/**
 * stopEncoder Function - Stop the Encoder given the camera namespace
 */
bool EncoderManager::stopEncoder(const std::string &camera_namespace)
{   
    encoder_status_[camera_namespace] = false;
    RCLCPP_INFO(this->get_logger(), "Publishing stopped for node %s", camera_namespace.c_str());
}   //  stopEncoder

}   //  namespace umrt_serial_cam_ros

File: umrt_serial_cam_ros/encoder_manager_node.py
"""
UMRT Encoder Manager Node Service

This Encoder Manager Node will start seperate services for each camera.
Currently we have 3 Serial Cameras, each named for their respective section,
Rover Cameras
Arm Cameras

TO-DO
- Convert this file to C++
- Launch Arguments in a seperate file to be pulled into this document
"""

"""
Import
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from rclpy.executors import MultiThreadedExecutor
from std_srvs.srv import SetBool
import subprocess
import os
import signal
import time
# from umrt_ros_msgs.srv import EncoderControl 
from functools import partial

"""
Encoder Manager Node Class
"""
class EncoderManager(Node):

    def __init__(self):
        super().__init__('encoder_manager')
        self.get_logger().info('Encoder Manager Node Started')

        # Store the Encoder Processes
        self.encoder_processes = {
            'rover_cam0': None,
            'arm_cam0': None,
            'arm_cam1': None
        }

        # Launch Arguments for different camera namespace
        self.encoder_launch_args = {
            'rover_cam0': {            
                'namespace': 'rover',
                'in_raw': '/rover/image_raw',
                'out_ffmpeg': '/rover/ffmpeg',
                'encoding': 'h264_nvenc',
                'qmax': '31',
                'preset': 'p1',
                'tune': 'll',
                'bit_rate': '2000000',
                'gop_size': '15'
            },

            'arm_cam0': {
                'namespace': 'arm0',
                'in_raw': '/arm0/image_raw',
                'out_ffmpeg': '/arm0/ffmpeg',
                'encoding': 'h264_nvenc',
                'qmax': '31',
                'preset': 'p1',
                'tune': 'll',
                'bit_rate': '2000000',
                'gop_size': '15'
            },

            'arm_cam1': {
                'namespace': 'arm1',
                'in_raw': '/arm1/image_raw',
                'out_ffmpeg': '/arm1/ffmpeg',
                'encoding': 'h264_nvenc',
                'qmax': '31',
                'preset': 'p1',
                'tune': 'll',
                'bit_rate': '2000000',
                'gop_size': '15'
            }
        }

        # self.srv = self.create_service(EncoderControl, 'encoder_control', self.encoder_control_callback)

        # Create Service for each camera
        #self.services = {}
        for cam in self.encoder_processes.keys():
            service_name = f'/{cam}/bool'

            # Create service
            self.create_service(
                SetBool,
                service_name,
                partial(self.encoder_control_callback, camera_namespace=cam)
            )
            
            self.get_logger().info(f"Created service: {service_name} for {cam}")

    def encoder_control_callback(self, request, response, camera_namespace: str):
        # camera_namespace = request.camera_namespace
        # command = request.command

        command_bool = request.data # Either True - start or False - stop
        command = "start" if command_bool else "stop"

        # # Check first if the Camera can be encoded
        # if camera_namespace not in self.encoder_processes:
        #     response.success = False
        #     response.message = f"Invalid camera name space: {camera_namespace}"
        #     self.get_logger().error(response.message)
        #     return response

        # Successful Camera namespace
        self.get_logger().info(f'Received command: {command} for {camera_namespace}')

        # Check if command is valid 
        # Start Command - To start the encoder
        if command == "start":
            # Check if encoder for that namespace is already running 
            # If it is running
            if self.encoder_processes[camera_namespace] is not None and self.encoder_processes[camera_namespace].poll is None:
                response.success = False
                response.message = f'Encoder {camera_namespace} is already running.'
                self.get_logger().warn(response.message)
            # Else if its not running 
            else:
                success, msg = self.start_encoder(camera_namespace)
                response.success = success
                response.message = msg
        # Stop Command - To stop the encoder
        elif command == "stop":
            # Check if encoder for that namespace has stopped 
            # If it is stopped
            if self.encoder_processes[camera_namespace] is None or self.encoder_processes[camera_namespace].poll() is not None:
                response.success = False
                response.message = f'Encoder {camera_namespace} is already stopped.'
                self.get_logger().warn(response.message)
            # Else if it not stopped
            else:
                success, msg = self.stop_encoder(camera_namespace)
                response.success = success
                response.message = msg
        # Invalid Command
        else:
            response.success = False
            response.message = f"Unknown command: {command}. Use 'start' or 'stop'"
            self.get_logger().error(response.message)

        return response
    
    """
    Start Encoder Function - It will start the encoder for a given camera namespace
    """
    def start_encoder(self, camera_namespace):
        # Try to start Encoder
        try:
            args = self.encoder_launch_args[camera_namespace]

            # Create the commmand with Launch Arguments
            cmd = ['ros2', 'launch', 'umrt-cam-launch-pkg', 'encode.launch.py']
            for key, value in args.items():
                cmd.append(f'{key}:={value}')

            self.get_logger().info(f"Starting encoder {camera_namespace}, with command: {' '.join(cmd)}")

            process = subprocess.Popen(cmd, preexec_fn=os.setsid)
            self.encoder_processes[camera_namespace] = process
            
            return True, f'Encoder {camera_namespace} started (PID: {process.pid})'
        
        # Failed to start Encoder
        except Exception as e:
            self.get_logger().error(f'Failed to start encoder {camera_namespace}: {e}')
            return False, f'Failed to start encoder {camera_namespace}: {e}'

    """
    Stop Encoder Function - It will stop the encoder for a given camera namespace
    """
    def stop_encoder(self, camera_namespace: str):
        # Get encoder process
        process = self.encoder_processes[camera_namespace]

        # Try if it is running
        if process:
            
            try:
                os.killpg(os.getpgid(process.pid), signal.SIGINT)
                process.wait(timeout=5)

                # Check if its still running even after the timeout
                if process.poll is None:
                    # Try SIGTERM
                    self.get_logger().warn(f'Encoder {camera_namespace} did not kill, sending SIGTERM')
                    os.killpg(os.getpgid(process.pid), signal.SIGTERM)
                    process.wait(timeout=2)
                    if process.poll is None:
                        # Try SIGKILL
                        self.get_logger().warn(f'Encoder {camera_namespace} did not kill, sending SIGKILL')
                        os.killpg(os.getpgid(process.pid), signal.SIGKILL)
                        process.wait(timeout=2)

                self.encoder_processes[camera_namespace] = None
                return True, f'Encoder {camera_namespace} stopped.'
            # Error that it already stopped or doesnt exist
            except ProcessLookupError:
                self.get_logger().warn(f'Encoder {camera_namespace} process not found')
                self.encoder_processes[camera_namespace] = None
                return True, f'Encoder {camera_namespace} already stopped or process not found'
            # Error with stopping the encoder
            except Exception as e:
                self.get_logger().error(f'Error stopping encoder: {camera_namespace}: {e}')
                return False, f'Error stopping encoder {camera_namespace}: {e}'
        # Else it is already stopped
        return False, f'Encoder {camera_namespace} is not running'
        
"""
Main
"""
def main(args=None):
    rclpy.init(args=args)
    encoder_manager = EncoderManager()

    executor = MultiThreadedExecutor()
    executor.add_node(encoder_manager)

    try:
        executor.spin()
    except KeyboardInterrupt:
        pass

    finally:
        # All subprocess are terminated
        for ns, process in encoder_manager.encoder_processes.items():
            if process and process.poll() is None:
                encoder_manager.get_logger().info(f'Terminating the remaining encoders for {ns}')
                # Terminate
                try:
                    os.killpg(os.getpgid(process.pid), signal.SIGTERM)
                    process.wait(timeout=5)
                except Exception as e:
                    encoder_manager.get_logger().error(f'Error during shutdown of {ns} encoder: {e}')
        
        # Kill Encoder Manager Node
        encoder_manager.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
